content/docs/api-reference/make_checkpoint.md

```diff
@@ -106,8 +103,7 @@ Experiment results have been applied to your workspace.
 
 In this example we kill the process (with Ctrl + C) after 3 checkpoints (at 0,
 100, and 200). The <abbr>cache</abbr> will contain those 3 versions of
-`int.txt`. DVC applies the last checkpoint to the <abbr>workspace</abbr> even if
-more cycles happened after that:
```

Comment:
Removed per https://github.com/iterative/dvc.org/pull/2286#discussion_r596867654 but I'm pretty sure I tried this and it was the case back then. Need to double check now... ⌛

Reply:
Hi @pmrowla,

So actually the last checkpoint is being applied to the workspace. In this case, int.txt contains `200` (or some other % 100 number). So did we get that wrong in docs after all (see link above) ? I'm on DVC 2.0.6.

Reply:
I'm on vacation and can't get into testing this, but looking at the example code I'm guessing it is because your "user" code does not cleanly handle the keyboard interrupt (and will exit with an error). If the user stage exits with an error, we don't do the extra step to generate another commit. We will generate the extra commit in situations like:

- User code runs a fixed number of checkpoint iterations and then does some other final step:

```python
for i in range(100):
    # do 100 iterations
    make_checkpoint()

# do something else that modifies the workspace
do_something()
```

- User code runs forever and is intended to be stopped (cleanly) with ctrl-c

```python
try:
    while True:
        # do something that modifies the workspace
        do_something()
        make_checkpoint()
except KeyboardInterrupt:
    pass
# modifications written in do_something() but interrupted before reaching
# make_checkpoint() will now be committed in DVC when this code returns/exits cleanly
```

I'm guessing that if you rewrite the docs example to work like this second case, you will see the non-100 values be written to your text file (and saved/committed in DVC)

Reply:
Basically, we generate a commit when a repro'd pipeline is reproduced cleanly. The potential "extra" commit at the end of checkpoint runs is just the same commit that is made for regular (non-checkpoint) `exp run`/`repro` runs.

When regular `exp run` or `repro` fails (because the stage/pipeline code exits with non-zero), we don't save anything, since we assume that the user does not want us to save the failed repo state. So if your checkpoint pipeline exits with a non-zero status, we will also not generate the extra/final commit.

Reply:
@jorgeorpinel I just tested and I think this particular example is causing confusion because `make_checkpoint()` takes way longer than the rest of the loop, making it almost impossible to interrupt at any point other than while `make_checkpoint()` is running. This creates the illusion that a checkpoint is being applied to the workspace. Add an `import time; time.sleep(0.1)` into the loop and you will be able to interrupt between checkpoints and see that the workspace reflects the current iteration, not the last checkpoint.

Reply:
> your example user code does not cleanly handle the keyboard interrupt
If the user stage exits with an error, we don't do the extra step to generate another commit

The example is simplified. The actual code I run catches the `KeyboardInterrupt`. The note about errors in interesting though, mentioned in eb029ecf

> the same commit that is made for regular (non-checkpoint) exp run/repro

My Q was about the application of the last regular checkpoint though (not the final extra checkpoint).
But again, the info about final checkpoints is def. valuable; I also added some of it (very briefly) to docs.
_Except that `repro` doesn't make commits. (?)_

Reply:
> this particular example is causing confusion because make_checkpoint() takes way longer than the rest of the loop...  This creates the illusion that a checkpoint is being applied to the workspace.

Ah that makes sense @dberenbaum ! OK that explains it, thanks. So the removal of this sentence is correct 👍 

Should I add that `time` stuff to this example for now? Probably not... I need to totally rewrite it soon anyway (not in this PR).

Reply:
Yeah, probably fine to leave it, especially if you are planning to replace it soon anyway.